runtime/permissions/lib.rs

```diff
@@ -310,12 +308,15 @@ impl AsRef<str> for EnvVarName {
   }
 }
 
-pub trait Descriptor: Eq + Clone + Hash {
-  type Arg: From<String>;
+pub trait QueryDescriptor: Debug {
+  type AllowDesc: AllowDescriptor<Query = Self>;
+  type DenyDesc: DenyDescriptor<Query = Self>;
```

Comment:
This is a bit confusing - effectively there are circular dependencies between these three types. Do you see any chance we could decouple that with a helper type?

Reply:
Also could you add some docstrings around these traits?

Reply:
I got rid of the additional traits and now there's only a QueryDescriptor.